Fix part of #2747, #1532: Remove support for Gradle

BenHenning opened the pull request:
## Explanation
Fixes part of #2747
Fixes #1532

This PR is doing a lot to address the vast majority of #2747--removing support for Gradle. It doesn't address everything, however--some items will need to be addressed in future PRs (largely new cleanup opportunities that are unlocked with the removal of Gradle). However, it does remove the entirety of Gradle definition infrastructure.

This PR specifically:
- Removes all Gradle build definitions (``.gradle`` files, ``settings.gradle``, the ``gradlew`` wrappers, and the ``gradle/wrapper`` directory). As a side benefit, this removes some of the last binary files from the codebase. :)
- Removes a bunch of Gradle-specific files not actually used by the Bazel builds, including proguard-rules.pro files (Bazel has separate definitions under ``config/proguard/``), some manifest files, and some Robolectric-specific definitions.
- Deletes the ``main.yml`` workflow (which was disabled in #5629). This is entirely a Gradle-specific workflow, so there's nothing that needs to be moved over as all of what's being tested is covered more efficiently and thoroughly in other specific workflow files.
- Updates to ``.gitignore`` per the Gradle file removals (fewer files/directories need to be ignored).
- Removed all TODOs on #1532 and marking it as fixed as part of this PR rather than performing the file rename. See https://github.com/oppia/oppia-android/issues/1532#issuecomment-2610933772 for an explanation on why.
- Some comments have been updated rather than addressed where it made sense that addressing the problem was not exactly specific to Gradle support (there's a desire to remove all references to 'Gradle' as a clear way to ensure everything Gradle-specific has been successfully removed from the codebase--this has largely been done).
- The infrastructure and definitions for segmenting tests between Bazel and Gradle has been completely removed (resulting in broad changes in the codebase) since Gradle-specific behaviors can't be tested anymore, and Bazel-specific behaviors are now the default.
- Some of the waiting logic in ``StateFragmentTest`` and ``ExplorationActivityTest`` has been kept but had its removal TODO removed. This is because the code seems to have changed (or the original problem misunderstood). These waits seem actually necessary due to relying on real image loading, so the fix is more complex than simply "migrate to Bazel & move off of Gradle." There's real test infrastructure work needed to fix these tests properly, and it doesn't seem sufficiently important to prioritize (though note it did seem possible to remove the waits outright in ``ExplorationActivityTest`` without issue). The longer-term infrastructure work is tracked in #1523.
- ``LoadLessonProtosFromAssets`` has been made default true (since it can't ever be false with Gradle removed). See the note below on removing JSON assets for the work to fully clean this up.
- The removal of ``OppiaTestRunner`` as it was only used for Gradle builds to help it set up state correctly for orchestrated tests (i.e. Espresso). We may need to reintroduce something like this in the future, but for now there's no value in keeping an unused class.
- Some simplifications in ``RepositoryFile`` and test removals in ``RepositoryFileTest`` now that Gradle build directories no longer need to be filtered.
- ``TestEnvironmentConfig`` was removed since there's no longer a need to test which environment tests are currently running in.
- Some Espresso documentation was removed since it's clearly no longer needed, but more documentation work is still needing to be completed (this was noticed due to the removal of ``OppiaTestRunner``).
- ``ApiMockLoader`` was simplified now that test config artifacts are in a single location during testing time.
- ``InitializeDefaultLocaleRule`` can now have nearly all of its reflection removed since it can be directly built to reference the classes that it's interacting with (which wasn't possible in the Gradle world, see the TODO in that class). Some new tests were added, as well, which brings its test coverage to 100%.
- Updates CODEOWNERS to reflect all of the many file changes from the above.
- Broad removal of the word 'module' in **many** locations and documentation. This term is generally replaced with the more architectural term 'layer' as to not overload 'module' from a Gradle, Bazel, and Dagger perspective (plus others). This required a lot of changes specifically in the Android Lint tool implementation and tests to try and keep the term 'module' specific to cases where it still makes sense (other than Dagger which is the bulk of the existing reference).
- `TransformAndroidManifest` was updated to support adding top-level comments in manifest files (which are TODOs for #6010 to move or remove manifest files--see below). Tests were also added to test the new functionality.
- A bunch of files were changed simply because my local editor removed trailing spaces and they were included in those various files that I happened to change (largely for 'module' reference renames and/or 'Gradle' reference removals/rewordings).

Note also that this work opens up more things that still need to be, or can now be, addressed, including:
- Removal of JSON lesson loading per #5663. We're no longer testing this behavior correctly as domain level tests would previously verify JSON loading logic in Gradle builds and proto logic in Bazel builds. This means there's a strong likelihood of breakages occurring if JSON loading isn't removed.
- Ensuring Espresso tests can actually run with Bazel (see #4991). We no longer can verify Espresso behaviors now that Gradle is removed, which also subsequently means all test gating on Robolectric is redundant and all tests gated on Espresso never run.
- It's possible #4748 is now made redundant with the removal of Gradle-specific ``robolectric.properties`` files, though I'm not entirely certain at this point which SDK version is used in Bazel tests for Robolectric (other than it being tied to Robolectric version and that we pull in all supported SDKs so that tests can selectively opt into specific SDKs for testing).
- Possibly some of the language configuration bits can be simplified now that we don't need to gracefully fail if the configuration is missing (since it never should be), but there's no plan to audit or track work for this.
- Reorganization of the codebase now that Gradle no longer imposes arbitrary modules (see https://github.com/oppia/oppia-android/issues/59#issuecomment-528217795) as well as fixing testing artifact isolation (multiple issues tracking parts of this, and ideas possibly not yet captured in issues).
- `genrule`s can now be removed to improve symbol analysis in Android Studio (and general build system complexity). See #1617.
- Some documentation can potentially be improved, but with the replacement of the terms 'Gradle' and 'module' (plus past documentation changes that happened as a result of moving the team over to Bazel exclusively) much of the existing documentation seems solid.
- It may now be possible to update to newer versions of Kotlin and Robolectric without worrying about AGP incompatibilities.
- Moving the app `AndroidManifest.xml` from the app directory to the top-level directory to address #1640. Unfortunately this can't be done as part of this PR (as previously attempted) due to #6010 and required changes to the Android Lint script. Note that the addressing of #6010 also allows other manifests to potentially be removed and, hopefully, a reduction of manifests to the bare minimum (the top-level plus any asset-holding libraries--at least for now).

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
There are no known UI effects from removing Gradle support, but build infrastructure is complex and this is _not_ a no-op change. It's expected that the Bazel builds are relatively the same as before, however (see the APK & AAB differences analysis), so there shouldn't be an UI or UX changes as a result of this work.

BenHenning committed:
First part in removing Gradle.

This removes the actual Gradle build infrastructure and CI workflows. It
also updates some TODO references to no longer reference Gradle (as
there are additional blockers for that work) and updates CODEOWNERS and
.gitignore accordingly. Some unnecesary files have also been removed.

BenHenning committed:
Remove .bazel suffix from all BUILD files.

This aligns with Bazel best practices and is preferred over the suffix
(which was only needed due to Gradle having some issues with files named
'BUILD'.

BenHenning committed:
Remove TODOs for renaming BUILD files.

BenHenning committed:
Remove build env test gating.

Bazel is now the only platform that these tests can run on. Note that no
tests were Gradle-specific.

BenHenning committed:
Remove TestEnvironmentConfig.

It is only used for checking if a test is running in Bazel, and all
tests are now.

One Gradle-specific test has been removed.

BenHenning committed:
Remove OppiaTestRunner.

This is only used when running tests with Gradle. We may eventually need
to reintroduce someting like this for Espresso support in the future,
but it's not yet clear if that's the case.

BenHenning committed:
Fix broken OppiaTestRul due to previous changes.

BenHenning committed:
Move primary AndroidManifest.xml to top-level.

Once modularization work is completed, this will be the only Android
manifest in the codebase. Since it contains high-level definitions of
the app structure, it makes sense to exist at the top level.

BenHenning committed:
Remove more Gradle-specific/caused workarounds.

BenHenning committed:
More removals and post clean-up fixes.

BenHenning committed:
Merge branch 'develop' into remove-gradle

BenHenning committed:
Address CI failures.

Specifically, this:
- Adds/removes CODEOWNER changes per file updates.
- Fixes two validation checks (which was confirmed).
- Removes redundant tests now that Gradle support has been removed.

BenHenning committed:
Undo .bazel extension rename.

It seems cleaner to retain the explicit extension in the long-term.

BenHenning committed:
Fix broken Buildifier script.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/walkthrough/topiclist/WalkthroughTopicListFragmentTest.kt:
```diff
@@ -218,38 +214,6 @@ class WalkthroughTopicListFragmentTest {
   }
 
-  @Test
-  fun testWalkthroughTopicListFragment_topicCard_lessonThumbnailIsCorrect() {
```
Try to re-add this. It seems wrong to remove it, actually.

BenHenning replied:
Tried enabling this in Bazel and am seeing the following failure:

```
1) testWalkthroughTopicListFragment_topicCard_lessonThumbnailIsCorrect(org.oppia.android.app.walkthrough.WalkthroughTopicListFragmentTest)
androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'with drawable from resource id: <2131230995>' doesn't match the selected view.
Expected: with drawable from resource id: <2131230995>[lesson_thumbnail_graphic_duck_and_chicken]
     Got: "LessonThumbnailImageView{id=2131297518, res-name=walkthrough_topic_thumbnail_image_view, visibility=VISIBLE, width=360, height=270, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.constraintlayout.widget.ConstraintLayout$LayoutParams@6361c171, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0}"

        at java.base/java.lang.Thread.getStackTrace(Thread.java:1602)
        at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:16)
        at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:36)
        at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:103)
        at androidx.test.espresso.ViewInteraction.check(ViewInteraction.java:31)
        at org.oppia.android.app.walkthrough.WalkthroughTopicListFragmentTest.testWalkthroughTopicListFragment_topicCard_lessonThumbnailIsCorrect(WalkthroughTopicListFragmentTest_updated.kt:237)
        ... 19 trimmed
Caused by: junit.framework.AssertionFailedError: 'with drawable from resource id: <2131230995>' doesn't match the selected view.
Expected: with drawable from resource id: <2131230995>[lesson_thumbnail_graphic_duck_and_chicken]
     Got: "LessonThumbnailImageView{id=2131297518, res-name=walkthrough_topic_thumbnail_image_view, visibility=VISIBLE, width=360, height=270, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.constraintlayout.widget.ConstraintLayout$LayoutParams@6361c171, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0}"

        at androidx.test.espresso.matcher.ViewMatchers.assertThat(ViewMatchers.java:17)
        at androidx.test.espresso.assertion.ViewAssertions$MatchesViewAssertion.check(ViewAssertions.java:15)
        at androidx.test.espresso.ViewInteraction$SingleExecutionViewAssertion.check(ViewInteraction.java:10)
        at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:11)
        at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:2)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at android.os.Handler.$$robo$$android_os_Handler$handleCallback(Handler.java:938)
        at android.os.Handler.handleCallback(Handler.java)
        at android.os.Handler.$$robo$$android_os_Handler$dispatchMessage(Handler.java:99)
        at android.os.Handler.dispatchMessage(Handler.java)
        at org.robolectric.shadows.ShadowPausedLooper$IdlingRunnable.run(ShadowPausedLooper.java:342)
        at org.robolectric.shadows.ShadowPausedLooper.executeOnLooper(ShadowPausedLooper.java:271)
        at org.robolectric.shadows.ShadowPausedLooper.idle(ShadowPausedLooper.java:89)
        at org.robolectric.android.internal.LocalControlledLooper.drainMainThreadUntilIdle(LocalControlledLooper.java:15)
        at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:99)
        ... 22 more
```

This may tie into #5663, actually, so I'm going to temporarily ignore the test and add a TODO on that issue to fix it.

BenHenning replied:
Actually, a better approach may be to just enable JSON loading for this test.

BenHenning replied:
Re-added in 6f064dc6a7ac67cf08abb4aca0e6c5285a8caea1.

BenHenning reviewed (commented):
Self-reviewed all changes.

BenHenning committed:
Re-add removed test & ensure it passes with Bazel.

BenHenning commented:
TODO: Consider removing references to 'module' as well since that's Gradle specific in most uses and may add confusion.

BenHenning committed:
Merge branch 'develop' into remove-gradle

BenHenning committed:
Merge branch 'develop' into remove-gradle

BenHenning committed:
Bring InitializeDefaultLocaleRule to 100% codecov.

BenHenning committed:
Remove Gradle 'module' terminology references.

Also, cleans up some Gradle-specific wiki stuff (more is needed yet).

BenHenning committed:
Rename oppia_android_module_level_test.

These are layer-level tests, not 'module' tests as that's a Gradle
concept that doesn't make sense in a Bazel-only world (and only adds
confusion).

BenHenning commented:
> TODO: Consider removing references to 'module' as well since that's Gradle specific in most uses and may add confusion.

Done. Also fixed the code coverage issue and brought this up-to-date with develop.

Still need to update the PR description with the latest changes, and finish cleaning up the wiki page changes. After those & any lingering CI problems are addressed, this PR should be ready for review.

BenHenning committed:
Merge branch 'develop' into remove-gradle

Conflicts:
	app/BUILD.bazel
	app/app_test.bzl
	app/src/main/java/org/oppia/android/app/testing/activity/BUILD.bazel
	app/test_with_resources.bzl
	scripts/assets/test_file_exemptions.textproto

BenHenning committed:
Merge branch 'develop' into remove-gradle

BenHenning committed:
Merge branch 'develop' into remove-gradle

BenHenning committed:
Merge branch 'develop' into remove-gradle

Conflicts:
	.github/workflows/main.yml
	gradle.properties
	wiki/Installing-Oppia-Android.md
	wiki/Oppia-Bazel-Setup-Instructions.md
	wiki/Troubleshooting-Installation.md

BenHenning committed:
Some post-merge clean-up.

Removes some trailing spaces and replaces 'module' references with
'layer' (when referring to Gradle). Also, remove some text that's no
longer relevant after this PR is merged (there will no longer be any
Gradle files in the codebase).

BenHenning committed:
Merge branch 'develop' into remove-gradle

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/test/resources/robolectric.properties
	data/src/main/AndroidManifest.xml
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationStorageModuleTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/testing/ExplorationStorageTestModuleTest.kt
	domain/src/test/resources/robolectric.properties
	testing/src/main/java/org/oppia/android/testing/network/ApiMockLoader.kt
	testing/src/main/java/org/oppia/android/testing/network/BUILD.bazel
	testing/src/test/resources/robolectric.properties

BenHenning committed:
Post-merge fixes.

BenHenning commented:
> > TODO: Consider removing references to 'module' as well since that's Gradle specific in most uses and may add confusion.
> 
> Done. Also fixed the code coverage issue and brought this up-to-date with develop.
> 
> Still need to update the PR description with the latest changes, and finish cleaning up the wiki page changes. After those & any lingering CI problems are addressed, this PR should be ready for review.

I need to recheck this now that I'm trying to bring this PR into review again.

Edit: Looks like the Android Lint tool will need to be updated, too, given that the manifest has moved locations.

BenHenning commented:
I started working on fixing the manifest issues. As far as I can tell there are two:
1. The app module manifest has moved to be top-level. This is an easy fix as it's just updating some path and test setup references and I was able to get beyond that failure in local changes.
2. In the case of the data 'module' its manifest is being removed since it isn't needed. As far as I can tell this is incompatible with the lint tool as it expects all Gradle 'modules' to have manifest files.

This begs a bigger question of actually continuing with this approach of simulating Gradle modules for linting when we're very explicitly moving away from that concept.

@manas-yu I'm curious if you have any suggestions how we should proceed here. Is there a way we can adapt the Android lint tool's design to work in a less "Gradle-module"-esque way?

manas-yu commented:
As discussed the best way forward considering the priority of current tasks would be to delay the refactoring of the Android lint script, and to unblock this PR and unmoving the Android Manifest files for now(This would be handled in the PR related to the refactoring of the Android Lint Script).

The Lint script is currently:
- Structured in a Gradle project structure(this was done to maintain/improve the accuracy of the lint reports as lint liked a Gradle line environment)
- Analyses the codebase based on modules app, data, domain, util, and testing and finds there manifest files at <module>/src/main/manifest.xml
- Utilises bazel cquery command in the BazelClient based on the target module tags

To migrate the android lint script to execute on a project level instead of depending on modules:
- Remove the dependency of the script on [Modules](https://github.com/oppia/oppia-android/blob/b799ed3db7009c67aded18a25a02a7930a2ebe6c/scripts/src/java/org/oppia/android/scripts/lint/LintCommon.kt#L11) 
- Treat the entire project as one large module passing in the src, resource, test files
- For AAR/JAR files use a bazel command like `bazel cquery "deps(//:project)" --output=files`
- Check for any hardcoded module based paths like <module>/src/main/manifest.xml and migrate them to a project level context
- Remove [Module Deps](https://github.com/oppia/oppia-android/blob/b799ed3db7009c67aded18a25a02a7930a2ebe6c/scripts/src/java/org/oppia/android/scripts/lint/LintProjectDescription.kt#L376) as they wont be needed.
- Make changes in the testing setup for the script for relevant testing.

Note: The refactoring can't guarantee similar accuracy to the current version as changes to the project description affects the reports majorly so the scripts accuracy would need to be tested again thoroughly.

BenHenning commented:
Thanks for that write-up @manas-yu! I've filed #6010 to track the tool-specific work here and will revert the changes that are breaking it.

BenHenning committed:
Revert manifest moves.

This ensures the Android lint tool now passes for the branch. It also
adds some new TODOs that needed to be supported in the manifest
transformer (so its tests were also updated accordingly).

BenHenning committed:
Fix module refs in Android lint script/tests/docs.

BenHenning committed:
Remove remaining Gradle references.

Revert some removed TODOs.

Revert the static analysis check for CI back to the way it was since the
manifest file was moved back.

BenHenning reviewed (commented):
Self-reviewed everything from past changes (more are still coming).

BenHenning committed:
Merge branch 'develop' into remove-gradle

BenHenning reviewed (commented):
Self-reviewed latest (quick pass).

BenHenning committed:
Update TODO exemptions.

BenHenning reviewed (commented):
Spot check exemption update.

BenHenning commented:
At this point I expect tests to pass unless they flake due to the ongoing zlib issue.

BenHenning commented:
PTAL for codeowners @seanlip and @adhiamboperes.

PTAL specifically for infrastructure, documentation, and lint changes @manas-yu.

@adhiamboperes and @manas-yu you are probably both broadly interested in all of the changes here since they are so far-reaching.

/cc @subhajitxyz as well for FYI.

seanlip commented on wiki/Overview-of-the-Oppia-Android-codebase-and-architecture.md:
```diff
@@ -14,75 +14,75 @@ The Oppia Android codebase makes use of Kotlin and XML and can be built using An
 
 
-The Oppia project follows a standard Gradle project layout with five modules. Each subdirectory in a module contains files that are specific to a particular feature or purpose.
+The Oppia project follows an approximate Gradle project layout with five layers. Each subdirectory in a layer contains files that are specific to a particular feature or purpose.
```
Do you still want the reference to Gradle here?

BenHenning replied:
Yep--I'm trying to communicate that the current project structure is roughly mimicking a Gradle project structure. I think this is the last or one of the last Gradle references left in the wiki after this PR.

seanlip commented on wiki/_Sidebar.md:
```diff
@@ -52,7 +52,7 @@
   * Bazel
     * [Exploring Bazel](https://github.com/oppia/oppia-android/wiki/Bazel-User-Guide)
-    * [Gradle Bazel Migration Best Practices and FAQ](https://github.com/oppia/oppia-android/wiki/Gradle--Bazel-Migration-Best-Practices-and-FAQ)
+    * [Bazel Migration Best Practices and FAQ](https://github.com/oppia/oppia-android/wiki/Gradle--Bazel-Migration-Best-Practices-and-FAQ)
```
Do you want to change the title of this page to not include Gradle?

BenHenning replied:
I really didn't want to rename the file, but I can.

BenHenning replied:
This has been done in the latest changes.

seanlip reviewed (approved):
Thanks @BenHenning! LGTM for codeowners, left 2 comments.

BenHenning committed:
Rename file to remove 'Gradle' from name.

adhiamboperes reviewed (commented):
Apologies for the dely @BenHenning. I should be able to complete this review later in the day. Things are however looking very meticulous.

adhiamboperes commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -4289,7 +4281,7 @@ test_file_exemption {
 test_file_exemption {
   exempted_file_path: "testing/src/main/java/org/oppia/android/testing/junit/InitializeDefaultLocaleRule.kt"
-  override_min_coverage_percent_required: 38
+  override_min_coverage_percent_required: 100
```
If we require full coverage for this file, shouldn't we just skip this override altogether like with the rest of the project?

BenHenning replied:
Technically this surpasses the minimum which is currently set to 70%:

https://github.com/oppia/oppia-android/blob/27d80bdeceefd28051ccfaf8b6139132b326c0fb/scripts/src/java/org/oppia/android/scripts/coverage/reporter/CoverageReporter.kt#L11

adhiamboperes commented on scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:
```diff
@@ -18,7 +18,7 @@ class RegexPatternValidationCheckTest {
   private val originalOut: PrintStream = System.out
   private val activitiesPlacementErrorMessage =
-    "Activities cannot be placed outside the app or testing module."
+    "Activities cannot be placed outside the top-level app or testing directories."
```
After thinking about this(per CLaM meeting discussion), I think this isn't as confusing as I earlier thought.

BenHenning replied:
Ack--we can definitely iterate on terminology if people do get confused moving forward though.

adhiamboperes commented on wiki/End-to-End-Testing-Guide.md:
```diff
@@ -10,12 +10,12 @@
 
 End-to-End tests in Oppia-android are written using [UiAutomator](https://developer.android.com/training/testing/ui-automator).
-These tests are written under the instrumentation module and don’t have Gradle support.
+These tests are written under the instrumentation directory.
```
Should this maybe be - 

```suggestion
These tests are written in a directory under the instrumentation layer.
```

BenHenning replied:
I don't think so. The instrumentation tests are not really a layer (at least not an app architectural layer which is what we've been using 'layer' to mean in this PR) which is actually why I kept it to directory here.

adhiamboperes reviewed (approved):
Finished reviewing this @BenHenning, nothing blocking. Generally LGTM.

Apologies for the delay!

BenHenning committed:
Merge branch 'develop' into remove-gradle

Conflicts:
	wiki/Installing-Oppia-Android.md

BenHenning commented:
I think I can address both of your comments @seanlip and @adhiamboperes, thanks for the reviews! Note that I didn't change anything from your comments @adhiamboperes (see my replies). Please let me know if there is something you'd like me to change--I'm happy to consider that for a follow-up PR as needed.

BenHenning reviewed (commented):
Checked merge commit & earlier reviewer address commit.

BenHenning commented:
This should be good to merge once CI is passing. Hopefully it won't take too many re-runs.
